[부산대 BE 이민형 - Step3] 상품 옵션

LeeSwallow opened the pull request:
### 특이사항
+ entity를 전반적으로 수정했습니다.
    + `BaseEntity`에 id가 포함되도록 설정했습니다.
    + `@Column` 을 통해 최소한의 제약조건을 명시했습니다.(nullable, unique ..)
    + option <-> product를 양방향 매핑하고 영속성을 보장하도록 구현하였습니다.

+ validator를 일부 수정했습니다.
    + `ValidtProductName` -> `validCharSet` + `Max` 로 두가지 일을 하는 문제를 일부 해결하고 `option`의 유효성검사에 사용했습니다.

+ option을 위한 엔드포인트를 만들었습니다.
    + `GET /products/{product_id}/options` (페이지네이션 조회)
    + `POST /products/{product_id}/options` (단건 옵션 생성)
    + `PUT /products/{product_id}/options/{option_id}` (단건 옵션 수정)
    + `PATCH /products/{product_id}/options/{option_id}` (단건 옵션 수량 조정, 멱등성 x)
    + `DELETE /products/{product_id}/options/{option_id}`(단건 옵션 삭제)

+ 이전에 구현하였던 aop(`PreAuthorize`)를 통해 사용자 권한 우선순위에 따라 user일 경우 option의 생성/수정/삭제는 상품의 소유자만 가능하도록 구현하였습니다. 

+ 상품 생성 페이지, 수정 페이지에 option을 생성할 수 있는 table를 추가로 구현했습니다. 

### 상품 상세 페이지
<img width="2024" height="1525" alt="image" src="https://github.com/user-attachments/assets/195e916e-2eb9-4ce3-aca4-7fdba20a58dc" />
<img width="925" height="462" alt="image" src="https://github.com/user-attachments/assets/18bda6dd-c01d-4407-bdac-f096ec697fec" />
> option을 추가 및 생성, 삭제 가능합니다. 
<img width="892" height="1097" alt="image" src="https://github.com/user-attachments/assets/1987b00c-e0cf-45ac-91a8-8ada14b3c3e9" />
> 제품 필드 또한 수정 가능합니다. 

### 상품 생성 페이지
<img width="853" height="918" alt="image" src="https://github.com/user-attachments/assets/5d92aa18-1594-4df1-bc73-2561c1141102" />
> 옵션을 추가적으로 넣을 수 있습니다.

### 유효성 검증 및 에러 핸들링

<img width="1197" height="1247" alt="image" src="https://github.com/user-attachments/assets/8411c643-c357-422a-9da3-5671dc3cc4e3" />

fetch api를 통해 각 요청을 보내고 실패 응답 시 validationError 및 errormessag를 추출해서 alert를 통해 문제사항을 알려주도록 간단하게 구현했습니다.

### 질문사항

+ 상품 생성 시 option을 무조건 넣어야되도록 강제하여 option이 1개는 있어야한다는 제약조건을 보장하였습니다. 이를 위해 상품 생성 요청 dto에 batch 방식으로 List<OptionRequest> 를 넣는 방식을 사용하였는데 이러한 배치 처리에 최대 크기를 설정하는 편이 좋을까요?

Commit:
refactor(entity) : BaseEntity id 추가 및 기본 entity 수정

- 중복되는 id 삭제
- 기본 생성자 protected로 변경
- entity 생성자를 체이닝으로 구성
- User의 Role은 Set으로 받아오도록 수정
- User의 SetRole 로직 수정

Commit:
fix(entity) : 변경사항 적용되도록 수정

Commit:
docs: 3 단계 구현 기능 목록 작성

Commit:
feat(entity): Option 엔티티 구현 & sql 테이블 추가

Commit:
feat(entity): OptionRepository & Service 구현

Commit:
refactor(mapper): EntityToDtoMapper 와 DtoToEntityMapper로 분리

Commit:
fix(service) : 필요 어노테이션 추가

Commit:
feat(dto): Option request / response dto 추가

- ValidProductName -> ValidCharSet 으로 변경하고 최대 글자를 확인하는 로직 삭제(단일 책임)
- 관련 어노테이션을 사용하는 dto에 @Size를 통한 문자열 크기 검증 추가
- create, update, patch request dto 추가 후 validation 구현

Commit:
feat(controller): OptionController 구현

- option service 관련 내부 로직 수정(수정 시 이름이 같으면 에러 반환)

LeeSwallow committed:
feat(database): sql의 제약조건을 수정

- entity에 nullable, unique 조건 명시 하도록 정리

LeeSwallow committed:
refactor(entity) : Product가 @ManyToOne으로 연관관계를 가져오도록 수정

- 확장성과 Action 관리를 위해 ownerId -> User owner로 변경
- 영향을 받은 코드 수정
- option uniquecontraint 추가

LeeSwallow committed:
feat(test) : option repository test 구현

LeeSwallow committed:
feat(test) : RestAssuredUtils 구현

- test 객체들을 생성하는 함수들을 미리 정의하여 중복 최소화
- 테스트 객체 생성 구문에 적용

LeeSwallow committed:
feat(test) : Option 생성(Create) E2E 테스트 구현

- AbstractOptionTest 추상 클래스로 테스트 객체 생성 전처리

LeeSwallow committed:
feat(test) : Option 읽기(Read) E2E 테스트 구현

- AbstractOptionTest 추상 클래스로 테스트 객체 생성 전처리

LeeSwallow committed:
feat(test) : Option 수정(Update) E2E 테스트 구현

LeeSwallow committed:
feat(test) : Option 수정(Delete) E2E 테스트 구현

LeeSwallow committed:
fix: option 도메인 관련 문제 수정

- OptionDefaultResponse에 createAt, updateAt 추가
- 관련 mapper 수정
- 잘못된 서비스 로직 수정(OptionServiceImpl)

LeeSwallow committed:
fix(entity) : 최소 한개의 옵션이 필요하다는 요구사항 반영

- List<Option>을 상품 엔티티에서 받아와 cascade.ALL 로 연관 엔티티를 함께 저장할 수 있도록 수정

LeeSwallow committed:
fix(dto) : 상품 생성 요청 수정

- 상품 생성과 동시에 옵션을 받아올 수 있도록 수정(옵션 필수)
- option을 생성 및 수정 시의 수량의 최소치를 1로 수정
- 옵션관련 validator 구현
- dto 관련 Mapper 수정

LeeSwallow committed:
fix: 옵션 수정 및 삭제 로직 수정

- 모든 옵션 삭제 제거(하나의 옵션이 필수 이므로)
- 옵션 수량 수정 결과가 0이 되지 못하도록 수정
- 옵션 제거 시 옵션의 수 확인 & 영속성 유지 하도록 수정

LeeSwallow committed:
fix: 관련 테스트 코드 수정 및 추가

Commit:
fix: 옵션 수량 최대치 설정(1억 미만)

- 관련 validation 추가
- 관련 service 예외 핸들링 추가
- 관련 e2e testcase 추가

Commit:
feat: 사용자 옵션을 추가할 수 있도록 설정

- fix: 엔티티 수정에 따라 ownerId -> owner.id로 소유자 ID를 가져올 수 있도록 수정
- options을 생성/삭제 가능하도록 수정
- fetch api 로 상품을 생성할 수 있도록 변경(options 반영)
- validationErrors를 가져와 에러를 출력할 수 있도록 구현

Commit:
feat: unit test에 영속성 테스트 추가

LeeSwallow committed:
feat : 상품 상세 및 수정 페이지에 option 추가

-fix : ProductUpdateRequest에 nullable이 isBlack에 의해 처리되지 않으므로 수정

urarik commented on src/main/java/gift/controller/api/UserController.java:
```diff
@@ -41,7 +42,7 @@ public ResponseEntity<CustomPage<UserAdminResponse>> getAllUsers(
         CustomPage<User> userPage = userService.findAllBy(pageable);
         return new ResponseEntity<>(
-                CustomPage.convert(userPage, EntityDtoMapper::toAdminDto), HttpStatus.OK
+                CustomPage.convert(userPage, EntityToDtoMapper::toAdminDto), HttpStatus.OK
```
메서드 레퍼런싱 좋네요 👍

urarik commented on src/main/java/gift/dto/option/OptionDefaultResponse.java:
```diff
@@ -0,0 +1,12 @@
+
+import java.time.Instant;
+
+public record OptionDefaultResponse(
```
Default가 아닌 Response도 있나요?

LeeSwallow replied:
UserAdminResponse, UserDefaultResonse 와 같이 권한수준에 따라 정보량이 다른 경우를 명시하고 싶었습니다. 수정하겠습니다.

urarik commented on src/main/java/gift/dto/option/PatchOptionRequest.java:
```diff
@@ -0,0 +1,20 @@
+public record PatchOptionRequest(
+    @Min(value = 1, message = "증감 시킬 수량은 1 이상이어야 합니다.")
+    Long quantity,
+    Boolean increment
```
그냥 quantity를 음수도 받을 수 있게 하면 increment 필드가 없어도 되지 않을까요?

urarik commented on src/main/java/gift/dto/option/PatchOptionRequest.java:
```diff
@@ -0,0 +1,20 @@
+        }
+        if (increment == null) {
+            increment = true;
+        }
```
이 값이 null이면 잘못된 요청인데 기본값을 넣어서 값을 증가시키는 것이 올바른 동작일까요?

urarik commented on src/main/java/gift/entity/Option.java:
```diff
@@ -0,0 +1,77 @@
+
+@Entity
+@Table(
+    name = "options",
```
테이블 이름을 복수로 한 이유가 있나요?

LeeSwallow replied:
테이블 명은 복수, 클래스는 단수형인 네이밍 규칙이 관례라고 생각했습니다.

urarik replied:
@LeeSwallow 그렇군여~ 저는 단수형이 관례라고 알고 있는데 혹시 민형님은 어디서 그런 관례를 보신건가요~?

LeeSwallow replied:
다른 프레임워크에서 복수형을 사용하는 것을 자주 봐왔고 이후에 "복수형을 쓰는 이유" 같은 방식으로 블로그 글을 참고 했던것 같습니다.  논쟁이 있는 사안이라고 생각하지 않고 사용하였는데 지적해주셔서 감사합니다.

urarik commented on src/main/java/gift/entity/Option.java:
```diff
@@ -0,0 +1,77 @@
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "product_id", nullable = false)
+    @OnDelete(action = OnDeleteAction.CASCADE)
```
Cascade 옵션 좋습니다 👍

urarik commented on src/main/java/gift/service/option/OptionService.java:
```diff
@@ -0,0 +1,16 @@
+import gift.entity.Option;
+import org.springframework.data.domain.Pageable;
+
+public interface OptionService {
```
인터페이스의 구현체가 1개인 것이 자명할 때 인터페이스를 추가하는 것이 어떤 이점이 있을까요?

LeeSwallow replied:
Mock 객체를 만드는 등에 테스팅에도 유용하게 사용할 수 있을것 같습니다.

urarik replied:
@LeeSwallow 꼭 인터페이스가 있어야만 목 객체를 만들 수 있을까요?

urarik commented on src/main/java/gift/service/option/OptionServiceImpl.java:
```diff
@@ -0,0 +1,140 @@
+    }
+
+
+    private Option updateAndReturn(Option option, String name, Long quantity) {
```
메소드 이름에 return이 들어간 이유가 있나요?

LeeSwallow replied:
wishlist 와 같이 quantity가 0이 되면 삭제하는 방향으로 updateOrDelete 를 구현하려다 노선을 틀게 되면서 불필요한 수식어가 붙게 된 것 같습니다

urarik commented on src/main/java/gift/service/option/OptionServiceImpl.java:
```diff
@@ -0,0 +1,140 @@
+        Product product = productService.findById(productId);
+        validateAuthorization(auth, product);
+        Option existingOption = findBy(id, productId);
+        return updateAndReturn(existingOption, null, existingOption.getQuantity() - quantity);
```
existingOption.decreaseQuantity(int amount)로 수정한 다음 바로 repository.save(exisitingOption)를 하는 것은 어떤까요? 제 생각엔 이 방식이 캡슐화의 관점에서도 더 좋아보입니다. quantity를 직접적으로 노출시키지 않으니깐요.

지금은 일반화된 케이스에서 쓰려고 updateAndReturn 메소드를 만든다음 이걸 계속 쓰려고 하다보니 불필요하게 복잡해진 것 같네요. 그리고 저라면 굳이 increase, decrease 메소드를 나눌 것 같지도 않아요.

urarik reviewed (changes requested):
수고하셨습니다. 코멘트 몇 개 달았으니 확인해주세요.

Commit:
refactor: dto 네이밍 규칙 변경

- reponse dto에 불필요한 default 삭제
- request는 {도메인}{동작}Request로 일관적으로 유지

Commit:
refactor: wishedProduct 도메인의 상품 증감 관련 service 및 dto 통합

- service 로직을 increase/decrease 구분없이 changeQauntityBy 로 수정
- dto로 증감할 수를 받아올때 increment 명시대신 음수값을 허용하도록 수정

Commit:
refactor: Option 도메인의 상품 증감 관련 service 및 dto 통합

- service 로직을 increase/decrease 구분없이 changeQauntityBy 로 수정
- dto로 증감할 수를 받아올때 increment 명시대신 음수값을 허용하도록 수정

Commit:
refactor : 모델 어트리뷰트로 페이지네이션 요청을 검증하도록 수정

LeeSwallow commented:
### 변경사항
+ increment / decrement 관련 로직과 dto를 change로 통합했습니다.
+ 페이지네이션를 모델 어트리뷰트로 받아오도록 변경했습니다.

urarik reviewed (approved):
한 주 수고 많으셨습니다~ 머지할게요. 코멘트 몇 개 달아놓았는데 확인해주세요!
그리고 추가로 해주신 커밋을 봤는데,
ProductRepositoryTest가 유닛 테스트일까요? 스프링 컨텍스트를 만드는 테스트가 유닛 테스트가 맞을까요~?
그리고 커밋에 refactor라고 붙여주신 것은 리팩토링을 의미하시는 것 같은데 리팩토링이 뭔기 한 번 공부해보세요~ 외부에 노출되는 API의 변경을 리팩토링이라고 할 수 있을까요?
